cpp/include/raft/cluster/kmeans_balanced.cuh

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <raft/cluster/detail/kmeans_balanced.cuh>
+#include <raft/core/mdarray.hpp>
+#include <raft/util/cuda_utils.cuh>
+
+namespace raft::cluster::kmeans_balanced {
+
+/**
+ * @brief Find clusters of balanced sizes with a hierarchical k-means algorithm.
+ *
+ * @code{.cpp}
+ *   #include <raft/core/handle.hpp>
+ *   #include <raft/cluster/kmeans_balanced.cuh>
+ *   #include <raft/cluster/kmeans_balanced_types.hpp>
+ *   ...
+ *   raft::handle_t handle;
+ *   raft::cluster::kmeans_balanced_params params;
+ *   auto centroids = raft::make_device_matrix<float, int>(handle, n_clusters, n_features);
+ *   raft::cluster::kmeans_balanced::fit(handle, params, X, centroids);
+ * @endcode
+ *
+ * @tparam DataT Type of the input data.
+ * @tparam MathT Type of the centroids and mapped data.
+ * @tparam IndexT Type used for indexing.
+ * @tparam MappingOpT Type of the mapping function.
+ * @param[in]  handle     The raft handle
+ * @param[in]  params     Structure containing the hyper-parameters
+ * @param[in]  X          Training instances to cluster. The data must be in row-major format.
+ *                        [dim = n_samples x n_features]
+ * @param[out] centroids  The generated centroids [dim = n_clusters x n_features]
+ * @param[in]  mapping_op (optional) Functor to convert from the input datatype to the arithmetic
+ *                        datatype. If DataT == MathT, this must be the identity.
+ */
+template <typename DataT, typename MathT, typename IndexT, typename MappingOpT = raft::identity_op>
+void fit(handle_t const& handle,
+         kmeans_balanced_params const& params,
+         raft::device_matrix_view<const DataT, IndexT> X,
+         raft::device_matrix_view<MathT, IndexT> centroids,
+         MappingOpT mapping_op = raft::identity_op())
+{
+  logger::get(RAFT_NAME).set_level(params.verbosity);
```

Comment:
Perhaps we should expect that the distance matches what is supported here too, and that k is a reasonable value (e.g. 0<k<n).